### ISSUE TITLE
[tune] Support flexible checkpointing in training functions and Trainables

### ISSUE THREAD

**Issue by ujvl** [ray-project/ray#6162]
Support some subset of: 
1. ~checkpoints to be taken mid-`_train` call~
2. checkpointing/restoring from within training functions (rather than just in `Trainable`)
3. more flexible checkpoint policies (rather than just passing in frequency). 

**Proposal** 

Function interface
```python
def train(config):
  if tune.track.can_load():
    model = restore_my_model(tune.track.load())  # restore from dict or path
  else:
    model = init_my_model()

  for i in range(start_iter, 100):
    ...
    result = {"acc": acc}
    if i % 10 == 0:
      # (1) save to files in dir
      checkpoint_dir = tune.track.get_checkpoint_dir()
      save_my_model(model, checkpoint_dir)
      # (2) OR save dict
      checkpoint = {"state": model.state()}
    
    tune.track.log(result, save=checkpoint)

tune.run(train)
```

Trainable interface
```python
class CheckpointPolicy:

  def should_checkpoint(result):
    """Returns whether or not checkpoint should be taken.
    Result is none for first call (before training begins).
    """
    raise NotImplementedError

  def checkpoint_score(result):
    """Returns the score of the checkpoint corresponding to the result."""
    raise NotImplementedError

tune.run(my_trainable, checkpoint_policy=PeriodicCheckpointPolicy(frequency=5))    
```

**Comment by AdamGleave** [comment]
Proposal LGTM. Calling `tune.track.{load,save}` is easy for me to add to code using Tune.

Is the intention that function interfaces use `tune.track.{load,save}` and `Trainable` continue with `_save()` and `_restore()`? What should happen if `tune.track.{load,save}` are called from within a `Trainable`?

What's the intended use of `CheckpointPolicy.checkpoint_score`? It seems `should_checkpoint` is in charge of checkpointing decisions, so presumably the score is purely informational. Is there a reason we'd want different scores for different checkpointing policies (as opposed to e.g. storing the loss which does not depend on checkpoint policy)?

**Comment by ujvl** [comment]
> Is the intention that function interfaces use tune.track.{load,save} and Trainable continue with _save() and _restore()?

Yep, analogous. 

> What should happen if tune.track.{load,save} are called from within a Trainable?

disallowed, will raise a clear error

> What's the intended use of CheckpointPolicy.checkpoint_score? 

used for checkpoint garbage collection; in the default policy this actually be `training_iteration` (ie GC the oldest checkpoints first)--consistent with current behaviour (see `checkpoint_score_attr`). but you can of course use something like accuracy or loss instead

the intention is just to consolidate a bunch of the existing parameters and pass in a single policy into `tune.run` rather than a bunch of things, eg:

```
policy = PeriodicCheckpointPolicy(frequency=4, 
                                  checkpoint_at_end=True, 
                                  scoring_attr="mean_loss",
                                  score_desc=True)
tune.run(..., checkpoint_policy=policy)
```

I realize this might be a bit unclear though, will document or refine if necessary

**Comment by AdamGleave** [comment]
Thanks for the clarification, those all seem like good decisions.

**Comment by richardliaw** [comment]
Ah, this is addressed with #8471
